driver/k8s/logicalvolume_service.go

```diff
@@ -221,6 +245,75 @@ func (s *LogicalVolumeService) DeleteVolume(ctx context.Context, volumeID string
 	}
 }
 
+// CreateSnapshot creates a snapshot of existing volume.
+func (s *LogicalVolumeService) CreateSnapshot(ctx context.Context, sourceVol *topolvmv1.LogicalVolume, sourceVolID, sname, snapType, accessType string) (string, error) {
+	logger.Info("CreateSnapshot called", "name", sname)
+	snapshotLV := &topolvmv1.LogicalVolume{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "LogicalVolume",
+			APIVersion: "topolvm.cybozu.com/v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: sname,
+		},
+		Spec: topolvmv1.LogicalVolumeSpec{
+			Name:        sname,
+			NodeName:    sourceVol.Spec.NodeName,
+			DeviceClass: sourceVol.Spec.DeviceClass,
+			Snapshot: topolvmv1.SnapshotSpec{
+				Type:       snapType,
+				DataSource: sourceVol.Status.VolumeID,
+				AccessType: accessType,
+			},
+		},
+	}
+
+	existingSnapshot := new(topolvmv1.LogicalVolume)
+	err := s.getter.Get(ctx, client.ObjectKey{Name: sname}, existingSnapshot)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return "", err
+		}
+		err := s.writer.Create(ctx, snapshotLV)
+		if err != nil {
+			return "", err
+		}
+		logger.Info("created LogicalVolume CR", "name", sname, "snapType", snapshotLV.Spec.Snapshot.Type, "datasource", snapshotLV.Spec.Snapshot.DataSource, "accessType", snapshotLV.Spec.Snapshot.AccessType)
+	} else {
+		if !existingSnapshot.IsCompatibleWith(snapshotLV) || sourceVolID != sourceVol.Status.VolumeID {
```

Comment:
check for `sourceVolID != sourceVol.Status.VolumeID` can be done early itself as its not dependent on the `Get` operation here?

Reply:
Since in this line, we are validating if the found existing volume is the right one or not; added both its checks in just one line to make it easier to understand and have checks for parent volume at the same place.
Let me know if it needs to be updated :+1:

Reply:
Since 
sourceVol, err := s.lvService.GetVolume(ctx, sourceVolID)

when can this ever be the case?

Reply:
I think the function does not need the sourceVol as an argument at all.